utils/wasm-gen/src/tests.rs

```diff
@@ -206,6 +213,120 @@ fn test_avoid_waits_works() {
     );
 }
 
+#[test]
+fn test_wait_stores_message_id() {
+    gear_utils::init_default_logger();
+
+    const EXPECTED_MSG_ID: u64 = 12345678;
+
+    let mut rng = SmallRng::seed_from_u64(123);
+    let mut buf = vec![0; UNSTRUCTURED_SIZE];
+    rng.fill_bytes(&mut buf);
+    let mut unstructured = Unstructured::new(&buf);
+
+    let params_config = SyscallsParamsConfig::new().with_default_regular_config();
+
+    let mut injection_types = SyscallsInjectionTypes::all_never();
+    injection_types.set(InvocableSyscall::Loose(SyscallName::Wait), 1, 1);
+    // Only for test, syscall `message_id` import added automatically when `wake` syscall is enabled in config.
+    injection_types.enable_syscall_import(InvocableSyscall::Loose(SyscallName::MessageId));
+
+    let syscalls_config_with_waiting_probabilty =
+        SyscallsConfigBuilder::new(injection_types.clone())
+            .with_params_config(params_config.clone())
+            .with_waiting_probability(NonZeroU32::new(4).unwrap())
+            .build();
+
+    let syscalls_config_wo_waiting_probabilty = SyscallsConfigBuilder::new(injection_types)
+        .with_params_config(params_config)
+        .build();
+
+    // Test both configs:
+    for syscalls_config in [
+        syscalls_config_with_waiting_probabilty,
+        syscalls_config_wo_waiting_probabilty,
+    ] {
+        let BackendReport {
+            termination_reason,
+            store,
+            memory,
+            ..
+        } = execute_wasm_with_custom_configs(
+            &mut unstructured,
+            syscalls_config,
+            ExecuteParams {
+                initial_memory_write: nonempty![set_init_called_flag(), set_wait_called_counter(0)]
+                    .into(),
+                message_id: EXPECTED_MSG_ID,
+                ..Default::default()
+            },
+        );
+
+        // It's Ok, message id is 32 bytes in size, but we use u64 for testing purposes.
+        let mut message_id = [0u8; 8];
+        let waited_message_id_ptr =
+            MemoryLayout::from(WASM_PAGE_SIZE * INITIAL_PAGES).waited_message_id_ptr;
+        memory
+            .read(&store, waited_message_id_ptr as u32, &mut message_id)
+            .unwrap();
+        assert_eq!(u64::from_le_bytes(message_id), EXPECTED_MSG_ID);
+
+        assert!(matches!(
+            termination_reason,
+            TerminationReason::Actor(ActorTerminationReason::Wait(..))
+        ));
+    }
+}
+
+#[test]
+fn test_wake_uses_stored_message_id() {
```

Comment:
As I see it, test first writes message id into memory and then calls `gr_wake`. I'm not sure if this is a good approach, because we want to check that syscall sequence `gr_message_id`, `gr_wake` causes `gr_wake` to use the last message id.